torch/_inductor/fx_passes/mkldnn_fusion.py

```diff
@@ -541,10 +544,298 @@ def _register_binary_unary_fusion():
+            return
+        if free_symbols(batch_size):
+            # TODO: support symbolic input size for bfloat16 path
+            return
+        else:
+            packed_weight_inputs = (mkldnn_tensor_node, batch_size)
+            packed_weight_op = (
+                torch.ops.mkldnn._reorder_linear_weight
+                if is_bf16_weight
+                else torch.ops.mkl._mkl_reorder_linear_weight
+            )
+            packed_weight_node = gm.graph.create_node(
+                "call_function", packed_weight_op, args=packed_weight_inputs
+            )
+        packed_linear_inputs = (input, packed_weight_node)
+        if is_bf16_weight:
+            packed_linear_inputs += (linear_bias, "none", [], "")
+            packed_linear_op = torch.ops.mkldnn._linear_pointwise.default
+        else:
+            packed_linear_inputs += (linear_weight, linear_bias, batch_size)
+            packed_linear_op = torch.ops.mkl._mkl_linear
+        packed_linear_node = gm.graph.create_node(
+            "call_function", packed_linear_op, packed_linear_inputs
+        )
+        linear_node.replace_all_uses_with(packed_linear_node)
+        packed_linear_node.meta.update(linear_node.meta)
+        gm.graph.erase_node(linear_node)
+        if len(permute_node.users) <= 1:
+            gm.graph.erase_node(permute_node)
+
+
+def _eliminate_duplicate_packed_nodes(gm):
+    """
+    Combine packed weight nodes with the same inputs to reduce memory usage.
+    for example:
+    class Model(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.linear = nn.Linear(32, 32, bias=True)
+
+        def forward(self, x):
+            return self.linear(self.linear(x))
+
+    the above's packed weight nodes are duplicate if two linear calls have same input size.
+    """
+    for node in gm.graph.nodes:
+        if node.target == "to_mkldnn" and len(node.args[0].users) > 1:
+            for user_node in list(node.args[0].users.keys()):
+                if user_node.target == "to_mkldnn" and user_node != node:
+                    user_node.replace_all_uses_with(node)
+                    gm.graph.erase_node(user_node)
+    packed_weight_ops = [
+        torch._C._nn.mkldnn_reorder_conv2d_weight,
+        torch.ops.mkldnn._reorder_convolution_transpose_weight,
+        torch.ops.mkldnn._reorder_linear_weight,
+        torch.ops.mkl._mkl_reorder_linear_weight,
+    ]
+    for node in gm.graph.nodes:
+        if node.target in packed_weight_ops and len(node.args[0].users) > 1:
+            for user_node in list(node.args[0].users.keys()):
+                if (
+                    user_node.target == node.target
+                    and user_node != node
+                    and user_node.args == node.args
+                ):
+                    user_node.replace_all_uses_with(node)
+                    gm.graph.erase_node(user_node)
+
+
+def mkldnn_weight_prepack_fx(gm, flat_params, fw_metadata):
```

Comment:
I need to land https://github.com/pytorch/pytorch/pull/100746, but I don't think we need this separate code path, or a lot of the changes.

As shown here:
https://github.com/pytorch/pytorch/pull/100746/files#diff-8ce7bc944676c03af78123e618ec440923b913f80d2c152db797071ff6318bc7R212-R218

we can register these pre-packing passes as a freezing pass. The `get_attr` [check](https://github.com/pytorch/pytorch/pull/100746/files#diff-4e47f07c143c95673f6877ea2d7d8778aa2222f20be3a603472a425c9ff76bbcR34-R37) is a way to check if the input is a constant, we don't need to thread constants around.

Reply:
@eellison, yes， register re-packing passes as a freezing passes.